src/components/Modal/Modal.tsx

```diff
@@ -0,0 +1,160 @@
+import React, { CSSProperties, useCallback, useEffect, useMemo } from 'react';
+import { CSSTransition } from 'react-transition-group';
+import classNames from 'classnames';
+import { ClassValue } from 'classnames/types';
+
+import ReactDOM from 'react-dom';
+import styles from './Modal.module.css';
+
+type Size = 'sm' | 'md' | 'lg';
+type Level = 'primary' | 'warning' | 'info' | 'danger';
+
+interface ModalProps {
+  children?: React.ReactNode;
+  className?: ClassValue;
+  isOpen?: boolean;
+  onClose?: () => void;
+  size?: Size;
+  style?: CSSProperties;
+}
+
+type ModalChildProps = Pick<ModalProps, 'children' | 'className'>;
+type ModalHeaderProps = Pick<ModalProps, 'children' | 'className'> & {
+  level?: Level;
+};
+
+type ModalType = React.FC<ModalProps> & {
+  Header: React.FC<ModalHeaderProps>;
+  Footer: React.FC<ModalChildProps>;
+  Body: React.FC<ModalChildProps>;
+};
+
+const Modal: ModalType = ({
+  children,
+  className: passedClassName = '',
+  isOpen = false,
+  onClose = () => {},
+  size = 'md',
+  style: passedStyle = {},
+  ...passedProps
+}) => {
+  const handleEscape = useCallback(
+    (e: KeyboardEvent) => {
```

Comment:
You should probably check to see if the cursor is focused on an element inside the modal before you call `onClose`

For instance, in your storybook implementation:
* Open the large modal (with the form)
* Start typing something
* Hit Escape
* I expect that the current input loses focus
* But actually, the modal closes (technically the input loses focus as well)

I think you can do this by getting the `document.activeElement` and a ref for your modal, then add a condition like
```
if (modalRef.current && !modalRef.current.contains(document.activeElement)) {
  onClose();
}
```